plugin/pkg/scheduler/equivalence_cache.go

```diff
@@ -61,74 +70,158 @@ func NewEquivalenceCache(getEquivalencePodFunc algorithm.GetEquivalencePodFunc)
+func (ec *EquivalenceCache) InvalidCachedPredicateItemForPod(nodeName string, predicateKeys sets.String, pod *api.Pod) {
+	if len(predicateKeys) == 0 {
+		return
+	}
 	ec.Lock()
 	defer ec.Unlock()
-	// clear cache of all nodes
-	for nodeName := range ec.algorithmCache {
-		delete(ec.algorithmCache, nodeName)
+	equivalenceHash := ec.getHashEquivalencePod(pod)
+	if equivalenceHash == 0 {
+		// no equivalence pod found, just return
+		return
+	}
+	for predicateKey := range predicateKeys {
+		if algorithmCache, exist := ec.algorithmCache[nodeName]; exist {
+			if cachePredicate, exist := algorithmCache.predicatesCache.Get(predicateKey); exist {
+				// got the cached item of by predicateKey & pod
+				predicateMap := cachePredicate.(PredicateMap)
+				if hostPredicate, ok := predicateMap[equivalenceHash]; ok {
+					// set the pod's item in predicateMap as invalid
+					hostPredicate.IsInvalid = true
+					// and add the predicateMap back to the cache
+					ec.algorithmCache[nodeName].predicatesCache.Add(predicateKey, predicateMap)
+				}
+			}
+		}
+	}
+	glog.V(5).Infof("Invalidate cached predicates %v on node %s, for pod %v", predicateKeys, nodeName, pod.GetName())
+}
+
+// InvalidCachedPredicateItemForPodAdd is a wrapper of InvalidCachedPredicateItem for pod add case
+func (ec *EquivalenceCache) InvalidCachedPredicateItemForPodAdd(pod *api.Pod, nodeName string) {
+	invalidPredicates := sets.NewString("GeneralPredicates")
+
+	// Invalid PodFitsHostPorts for all pods because we don't know if other pods require same HostPort
+	// `if len(predicates.GetUsedPorts(pod)) != 0` we should also invalidate "PodFitsHostPorts", but it has been
+	// included in "GeneralPredicates"
+
+	// If a pod scheduled to a node with these PV, it may cause disk conflict.
+	for _, volume := range pod.Spec.Volumes {
+		if volume.GCEPersistentDisk != nil || volume.AWSElasticBlockStore != nil || volume.RBD != nil {
+			invalidPredicates.Insert("NoDiskConflict")
+		}
+	}
+	ec.InvalidCachedPredicateItem(nodeName, invalidPredicates)
+
+	// MatchInterPodAffinity, this case is complicated, we'll invalid equivalence cache of all nodes
+	affinity, err := api.GetAffinityFromPodAnnotations(pod.Annotations)
+	if err != nil {
+		glog.Errorf("Failed to get affinity from pod annotation for equivalence cache")
+	}
+	if affinity != nil {
+		ec.InvalidCachedPredicateItemOfAllNodes(sets.NewString("MatchInterPodAffinity"))
 	}
 }
 
-// hashEquivalencePod returns the hash of equivalence pod.
-func (ec *EquivalenceCache) hashEquivalencePod(pod *v1.Pod) uint64 {
+// getHashEquivalencePod returns the hash of equivalence pod.
+// if no equivalence pod found, return false
+func (ec *EquivalenceCache) getHashEquivalencePod(pod *api.Pod) uint64 {
 	equivalencePod := ec.getEquivalencePod(pod)
-	hash := adler32.New()
-	hashutil.DeepHashObject(hash, equivalencePod)
-	return uint64(hash.Sum32())
+	if equivalencePod != nil {
+		hash := adler32.New()
```

Comment:
Opened #40140 since it turns out there are a couple of more places where adler is used.